File: android/app/src/main/java/com/stablechannels/app/util/TestOverrides.kt
package com.stablechannels.app.util

import android.content.Context
import android.util.Log
import com.stablechannels.app.BuildConfig
import org.json.JSONObject
import java.io.File

/**
 * Debug-only endpoint overrides for E2E testing against the regtest harness
 * (e2e/harness). Release builds NEVER read the file (BuildConfig.DEBUG gate).
 *
 * A test rig pushes a JSON file after the harness boots (the LSP node id is
 * only known then — that's why this is runtime, not build config):
 *
 *   adb push test_config.json \
 *     /sdcard/Android/data/com.stablechannels.app/files/test_config.json
 *
 * Shape (all keys optional; absent = production value):
 * {
 *   "network": "regtest",
 *   "primary_chain_url": "http://10.0.2.2:30000",
 *   "fallback_chain_url": "http://10.0.2.2:30000",
 *   "lsp_pubkey": "<ldk-server node id>",
 *   "lsp_address": "10.0.2.2:9735",
 *   "push_register_url": "http://10.0.2.2:9737/register-push",
 *   "channel_exists_url": "http://10.0.2.2:9737/channel-exists",
 *   "price_feed_base": "http://10.0.2.2:9737"
 * }
 *
 * "price_feed_base" replaces ALL five price feeds with the harness's
 * /feeds/<name> endpoints (same JSON shapes as the real feeds).
 */
object TestOverrides {
    private const val TAG = "TestOverrides"
    private const val FILE_NAME = "test_config.json"

    @Volatile private var initialized = false

    @Volatile var network: String? = null; private set
    @Volatile var primaryChainUrl: String? = null; private set
    @Volatile var fallbackChainUrl: String? = null; private set
    @Volatile var lspPubkey: String? = null; private set
    @Volatile var lspAddress: String? = null; private set
    @Volatile var pushRegisterUrl: String? = null; private set
    @Volatile var channelExistsUrl: String? = null; private set
    @Volatile var priceFeedBase: String? = null; private set

    val active: Boolean get() = network != null || lspPubkey != null || primaryChainUrl != null

    fun init(context: Context) {
        if (initialized) return
        initialized = true
        if (!BuildConfig.DEBUG) return
        try {
            val f = File(context.getExternalFilesDir(null), FILE_NAME)
            if (!f.exists()) return
            val json = JSONObject(f.readText())
            fun opt(key: String) = json.optString(key).takeIf { it.isNotBlank() }
            network = opt("network")
            primaryChainUrl = opt("primary_chain_url")
            fallbackChainUrl = opt("fallback_chain_url")
            lspPubkey = opt("lsp_pubkey")
            lspAddress = opt("lsp_address")
            pushRegisterUrl = opt("push_register_url")
            channelExistsUrl = opt("channel_exists_url")
            priceFeedBase = opt("price_feed_base")
            Log.w(TAG, "E2E overrides ACTIVE: network=$network lsp=$lspAddress chain=$primaryChainUrl")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load $FILE_NAME — using production endpoints", e)
        }
    }

    /** Harness-served replacements for the five production price feeds. */
    fun priceFeeds(base: String): List<PriceFeedConfig> = listOf(
        PriceFeedConfig("Bitstamp", "$base/feeds/bitstamp", listOf("last")),
        PriceFeedConfig("CoinGecko", "$base/feeds/coingecko", listOf("bitcoin", "usd")),
        PriceFeedConfig("Kraken", "$base/feeds/kraken", listOf("result", "XXBTZUSD", "c")),
        PriceFeedConfig("Coinbase", "$base/feeds/coinbase", listOf("data", "amount")),
        PriceFeedConfig("Blockchain.com", "$base/feeds/blockchain", listOf("USD", "last"))
    )
}

File: ios/StableChannels/StableChannels/Utilities/TestOverrides.swift
import Foundation

/// Debug-only endpoint overrides for E2E testing against the regtest harness
/// (e2e/harness). Release builds NEVER read the file (compile-time gate).
///
/// The test rig writes test_config.json into the app's Documents directory
/// (simulator: `xcrun simctl get_app_container booted <bundle> data`) after
/// the harness boots — the LSP node id is only known then, which is why this
/// is a runtime file, not build configuration. Same JSON shape as Android's
/// TestOverrides (util/TestOverrides.kt).
enum TestOverrides {
    struct Values {
        var network: String?
        var primaryChainUrl: String?
        var fallbackChainUrl: String?
        var lspPubkey: String?
        var lspAddress: String?
        var pushRegisterUrl: String?
        var channelExistsUrl: String?
        var priceFeedBase: String?
        var disableSendAuth = false
        var syncIntervalSecs: UInt64?
    }

    /// Loaded once on first access; thread-safe via `static let`.
    static let shared: Values = load()

    private static func load() -> Values {
        var v = Values()
        #if DEBUG
            let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
            guard let url = docs?.appendingPathComponent("test_config.json"),
                  let data = try? Data(contentsOf: url),
                  let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
            else { return v }
            func opt(_ key: String) -> String? {
                (json[key] as? String).flatMap { $0.isEmpty ? nil : $0 }
            }
            v.network = opt("network")
            v.primaryChainUrl = opt("primary_chain_url")
            v.fallbackChainUrl = opt("fallback_chain_url")
            v.lspPubkey = opt("lsp_pubkey")
            v.lspAddress = opt("lsp_address")
            v.pushRegisterUrl = opt("push_register_url")
            v.channelExistsUrl = opt("channel_exists_url")
            v.priceFeedBase = opt("price_feed_base")
            v.disableSendAuth = (json["disable_send_auth"] as? Bool) ?? false
            if let secs = (json["sync_interval_secs"] as? NSNumber)?.uint64Value, secs > 0 {
                v.syncIntervalSecs = secs
            }
            print(
                "[TestOverrides] E2E overrides ACTIVE: network=\(v.network ?? "-") lsp=\(v.lspAddress ?? "-") chain=\(v.primaryChainUrl ?? "-")"
            )
        #endif
        return v
    }

    /// Harness-served replacements for the five production price feeds.
    static func priceFeeds(base: String) -> [PriceFeedConfig] {
        [
            PriceFeedConfig(name: "Bitstamp", urlFormat: "\(base)/feeds/bitstamp", jsonPath: ["last"]),
            PriceFeedConfig(name: "CoinGecko", urlFormat: "\(base)/feeds/coingecko", jsonPath: ["bitcoin", "usd"]),
            PriceFeedConfig(name: "Kraken", urlFormat: "\(base)/feeds/kraken", jsonPath: ["result", "XXBTZUSD", "c"]),
            PriceFeedConfig(name: "Coinbase", urlFormat: "\(base)/feeds/coinbase", jsonPath: ["data", "amount"]),
            PriceFeedConfig(name: "Blockchain.com", urlFormat: "\(base)/feeds/blockchain", jsonPath: ["USD", "last"])
        ]
    }
}
